File: src/experiments/genetics_rust/js/logic/LivingNode.ts
import NodeLogic from "./NodeLogic";

function clamp(val: number, min: number, max: number) {
	return Math.max(Math.min(val, max), min);
}

class LivingNode {
	health: number;
	pos: {x: number, y: number};
	vel: {x: number, y: number};
	logic: NodeLogic;
	reward: number;
	distanceMoved: number;
	bounds: { x: number; y: number };

	constructor(
		logic: NodeLogic,
		bounds: { x: number; y: number },
		pos: { x: number; y: number }
	) {
		this.health = 1400;
		this.pos = pos;
		this.vel = { x: 0, y: 0 };
		this.logic = logic;
		this.reward = 0;
		this.bounds = bounds;
		this.distanceMoved = 0;
	}

	async tick(targetPos: { x: number; y: number }, newTargetTimeout: number, ticksSinceNewTarget: number) {
		if (this.isDead()) return;
		let dx = this.pos.x - targetPos.x;
		let dy = this.pos.y - targetPos.y;
		const action = this.logic.step(new Float32Array([
			this.pos.x / (this.bounds.x * 0.5) - 1, this.pos.y / (this.bounds.y * 0.5) - 1,
			this.vel.x, this.vel.y,
			targetPos.x / (this.bounds.x * 0.5) - 1, targetPos.y / (this.bounds.y * 0.5) - 1/*,
			newTargetTimeout * 2 - 1*/
		]));
		//console.log(action[0], " ", action[1]);
		// action is an array with 2 values: x and y movement
		let dist = Math.sqrt(action[0] * action[0] + action[1] * action[1]);
		//this.distanceMoved += dist;

		if (dist > 1) {
			action[0] /= dist;
			action[1] /= dist;
		}
		this.vel.x += clamp(action[0], -1, 1) * 0.02;
		this.vel.y += clamp(action[1], -1, 1) * 0.02;
		const velMag = Math.sqrt(this.vel.x * this.vel.x + this.vel.y + this.vel.y);
		if(velMag > 1){
			this.vel.x /= velMag;
			this.vel.y /= velMag;
		}
		const prevPos = {x : this.pos.x, y: this.pos.y };
		this.pos.x += this.vel.x;
		this.pos.y += this.vel.y;
		this.snapBounds(prevPos);

		if (this.health % 10 === 0 && ticksSinceNewTarget > 30) {
			let dist = Math.sqrt(dx * dx + dy * dy);
			let maxDist = Math.sqrt(
				this.bounds.x * this.bounds.x + this.bounds.y * this.bounds.y
			);
			/*let dist = Math.abs(dx) + Math.abs(dy);
			let maxDist = Math.abs(this.bounds.x) + Math.abs(this.bounds.y);*/

			this.reward += 1 - dist / maxDist;
			if(dist < maxDist * 0.05)
				this.reward += 0.5;
			
		}
		this.health -= 1; // loose 1 health per tick
		//this.health -= dist; // loose health proportional to movement
	}

	getFitness() {
		return this.reward;
	}

	endEpisode() {
		//if (this.distanceMoved < 10) this.reward -= 1000;
	}

	isDead() {
		return this.health <= 0;
	}

	snapBounds(prevPos: { x: number, y: number }) {
		// Line in the middle that constricts to center
		const sizeOfBarrier = 0.1 * 0.5 *  this.bounds.x;
		const halfXBound = this.bounds.x / 2;
		if((this.pos.x > halfXBound - sizeOfBarrier && prevPos.x <= halfXBound - sizeOfBarrier) || 
			(this.pos.x < halfXBound + sizeOfBarrier && prevPos.x >= halfXBound + sizeOfBarrier)){
				// crossed the border via x
				if(this.pos.y > this.bounds.y * 0.2 && this.pos.y < this.bounds.y * 0.8){
					
					if(prevPos.x > halfXBound)
						this.pos.x = halfXBound + sizeOfBarrier;
					else
						this.pos.x = halfXBound - sizeOfBarrier;
					this.vel.x = 0;
				}
			}
		if((this.pos.y < this.bounds.y * 0.8 && prevPos.y >= this.bounds.y * 0.8) || 
			(this.pos.y > this.bounds.y * 0.2 && prevPos.y <= this.bounds.y * 0.2)){
				if(this.pos.x >= halfXBound - sizeOfBarrier && this.pos.x <= halfXBound + sizeOfBarrier){
					
					if(prevPos.y > this.bounds.y * 0.5)
						this.pos.y = this.bounds.y * 0.8;
					else
						this.pos.y = this.bounds.y * 0.2;
					this.vel.y = 0;
				}
			}

		// Outer border
		if (this.pos.x > this.bounds.x) {
			this.pos.x = this.bounds.x;
			this.vel.x = 0;
		} else if (this.pos.x < 0) {
			this.pos.x = 0;
			this.vel.x = 0;
		}

		if (this.pos.y > this.bounds.y) {
			this.pos.y = this.bounds.y;
			this.vel.y = 0;
		} else if (this.pos.y < 0) {
			this.pos.y = 0;
			this.vel.y = 0;
		}
	}
}

export default LivingNode;

File: src/experiments/genetics_rust/wasm/src/living_node.rs
use crate::node_logic::{NodeLogic};
use crate::vec::Vec2;

#[derive(Clone)]
pub struct LivingNode {
    health: i32,
    pub pos: Vec2,
    vel: Vec2,
    pub logic: NodeLogic,
    reward: f64,
    bounds: Vec2,
}

impl LivingNode {
    pub fn new(logic: NodeLogic, bounds: Vec2, pos: Vec2) -> LivingNode {
        LivingNode {
            health: 2000,
            pos,
            vel: Vec2 { x: 0.0, y: 0.0 },
            logic,
            reward: 0.0,
            bounds
        }
    }

    #[inline(always)]
    pub fn is_dead(&self) -> bool {
        self.health <= 0
    }

    fn snap_bounds(&mut self, prev_pos: &Vec2) {
        let size_x = 0.1 * 0.5 * self.bounds.x;
        let half_x_bound = self.bounds.x / 2.0;

        if (self.pos.x > half_x_bound - size_x && prev_pos.x <= half_x_bound - size_x) ||
            (self.pos.x < half_x_bound + size_x && prev_pos.x >= half_x_bound + size_x) {
            // crossed the border via x
            if self.pos.y > self.bounds.y * 0.2 && self.pos.y < self.bounds.y * 0.8 {
                if prev_pos.x > half_x_bound {
                    self.pos.x = half_x_bound + size_x;
                } else {
                    self.pos.x = half_x_bound - size_x;
                }
                self.vel.x = 0.0;
            }
        }

        if (self.pos.y < self.bounds.y * 0.8 && prev_pos.y >= self.bounds.y * 0.8) ||
            (self.pos.y > self.bounds.y * 0.2 && prev_pos.y <= self.bounds.y * 0.2) {
            // crossed the border via y
            if self.pos.x >= half_x_bound - size_x && self.pos.x <= half_x_bound + size_x {
                if prev_pos.y > self.bounds.y * 0.5 {
                    self.pos.y = self.bounds.y * 0.8;
                } else {
                    self.pos.y = self.bounds.y * 0.2;
                }
                self.vel.y = 0.0;
            }
        }

        // outer border
        if self.pos.x > self.bounds.x {
            self.pos.x = self.bounds.x;
            self.vel.x = 0.0;
        } else if self.pos.x < 0.0 {
            self.pos.x = 0.0;
            self.vel.x = 0.0;
        }

        if self.pos.y > self.bounds.y {
            self.pos.y = self.bounds.y;
            self.vel.y = 0.0;
        } else if self.pos.y < 0.0 {
            self.pos.y = 0.0;
            self.vel.y = 0.0;
        }
    }

    #[inline(always)]
    pub fn get_fitness(&self) -> f64 {
        self.reward
    }

    pub fn get_next_action(&mut self, target_pos: Vec2) -> Vec2 {
        let mut action = Vec2::from_slice(self.logic.step(&[
            self.pos.x / (self.bounds.x * 0.5) - 1.0,
             self.pos.y / (self.bounds.y * 0.5) - 1.0,
             self.vel.x,
             self.vel.y,
             target_pos.x / (self.bounds.x * 0.5) - 1.0,
             target_pos.y / (self.bounds.y * 0.5) - 1.0
         ]));

        let action_mag = action.length();
        if action_mag > 1.0 {
            action._div(action_mag);
        }

        action
    }

    pub fn tick(&mut self, target_pos: Vec2, action: Vec2, ticks_since_new_target: i32) {
        if self.is_dead() {
            return;
        }

        self.vel.x += action.x.clamp(-1.0, 1.0) * 0.02;
        self.vel.y += action.y.clamp(-1.0, 1.0) * 0.02;

        let vel_len = self.vel.length();
        if vel_len > 1.0 {
            self.vel._div(vel_len);
        }

        let prev_pos = self.pos;
        self.pos._add(&self.vel);
        self.snap_bounds(&prev_pos);

        if self.health % 10 == 0 && ticks_since_new_target > 250 {
            let dist = self.pos.sub(&target_pos).length();
            let max_dist = self.bounds.length();

            self.reward += 1.0 - dist / max_dist;
        }

        self.health -= 1;
    }
}